src/System.Linq/tests/SelectTests.cs

```diff
@@ -0,0 +1,230 @@
+
+            IEnumerable<int> query = source.Select(d => d).Select(d => d());
+            Assert.False(funcCalled);
+        }
+
+        [Fact]
+        public void SelectSelect_SourceIsIEnumerable_ExecutionIsDefered()
+        {
+            bool funcCalled = false;
+            IEnumerable<Func<int>> source = Enumerable.Repeat((Func<int>)(() => { funcCalled = true; return 1; }), 1);
+
+            IEnumerable<int> query = source.Select(d => d).Select(d => d());
+            Assert.False(funcCalled);
+        }
+
+        [Fact]
+        public void SelectSelect_SourceIsAnArray_ReturnsExpectedValues()
+        {
+            Func<int, int> selector = i => i + 1;
+            int[] source = new[] { 1, 2, 3, 4, 5 };
+
+            IEnumerable<int> query = source.Select(selector).Select(selector);
+
+            int index = 0;
+            foreach (var item in query)
+            {
+                var expected = selector(selector(source[index]));
+                Assert.Equal(expected, item);
+                index++;
+            }
+
+            Assert.Equal(source.Length, index);
+        }
+
+        [Fact]
+        public void SelectSelect_SourceIsAList_ReturnsExpectedValues()
+        {
+            List<int> source = new List<int> { 1, 2, 3, 4, 5 };
+            Func<int, int> selector = i => i + 1;
+
+            IEnumerable<int> query = source.Select(selector).Select(selector);
+
+            int index = 0;
+            foreach (var item in query)
+            {
+                var expected = selector(selector(source[index]));
+                Assert.Equal(expected, item);
+                index++;
+            }
+
+            Assert.Equal(source.Count, index);
+        }
+
+        [Fact]
+        public void SelectSelect_SourceIsIEnumerable_ReturnsExpectedValues()
+        {
+            int nbOfItems = 5;
+            IEnumerable<int> source = Enumerable.Range(1, 5);
+            Func<int, int> selector = i => i + 1;
+
+            IEnumerable<int> query = source.Select(selector).Select(selector);
+
+            int index = 0;
+            foreach (var item in query)
+            {
+                index++;
+                var expected = selector(selector(index));
+                Assert.Equal(expected, item);
+            }
+
+            Assert.Equal(nbOfItems, index);
+        }
+
+        [Fact]
+        public void Select_GetEnumeratorCalledTwice_DifferentInstancesReturned()
+        {
+            int[] source = new[] { 1, 2, 3, 4, 5 };
+            var query = source.Select(i => i + 1);
+
+            var enumerator1 = query.GetEnumerator();
+            var enumerator2 = query.GetEnumerator();
+
+            Assert.NotSame(enumerator1, enumerator2);
+
+            enumerator1.Dispose();
+            enumerator2.Dispose();
+        }
+
+        private void AssertException<TException>(Action action) where TException : Exception
+        {
+            try
+            {
+                action();
+            }
```

Comment:
AFAIK test will fail if no assert is executed, but yeah, I'm changing to `Assert.Throws` and adding *System.Threading.Tasks* reference.